### PR TITLE
Add support for multi phase OCP in Python

### PR DESCRIPTION
**Pull request by FreyJo**
Multi phase OCPs can be formulated using the `AcadosMultiphaseOcp` class.
An `AcadosOcpSolver` can be created from an `AcadosMultiphaseOcp` object, using:
```
ocp_solver = AcadosOcpSolver(mocp)
```
Interactions with the solver are the same as when it is created from an `AcadosOcp` formulation object.

#### Examples and tests using multi-phase OCPs
`examples/acados_python/pendulum_on_cart/ocp/ocp_example_cost_formulations.py`
`examples/acados_python/time_varying/piecewiese_polynomial_control_example.py`
`examples/acados_python/time_varying/test_mocp_qp.py`

### Formulating a multi-phase OCP
Is intended to be done as follows:
```
N_horizon = sum(N_list)
n_phases = len(N_list)

mocp = AcadosMultiphaseOcp(N_list)

for i in range(n_phases):
    ocp = create_ocp_formulation_without_opts(...)
    mocp.set_phase(ocp, i)

# set options
mocp.solver_options.qp_solver = 'PARTIAL_CONDENSING_HPIPM'
...
```
Here, `n_phases` is the number of phases, `N_list` a list containing the number of shooting intervals corresponding to each phase.
Each phase is first formulated as an `AcadosOcp` object and then set in the `mocp` formulation object.

Internally, `AcadosMultiphaseOcp` contains arrays of length `n_phases` for model, cost, and constraints (`AcadosModel`, `AcadosOcpCost`, `AcadosOcpConstraints`).
Initial cost and constraints are set by the first phase. If such cost or constraints, i.e. fields ending with `_0`, are defined by later phases, these are ignored, after warning the user about it.
Similarly, terminal cost and constraints are defined only by the last phase.

The class only contains a single `AcadosOcpOptions` object, since most options are not phase dependent, exceptions see limitations below, which will be addressed in a follow up PR.

### Current limitations of the implementation:
- state dimension must be the same for all phases (all other dimensions can be different), extending would require defining a transition function.
    - solver options are shared between all phases. Interesting options to set phase-wise are:
        - `integrator_type`
        - `collocation_type`
        - `cost_discretization`

### CODE REVIEW

**Comment by FreyJo** [comment]
One suggestion is to move the definition of `N` to `AcadosOcpOptions`.
Notes:
- no dimensions would need to be set
- it is more of an option and closely connected to e.g. `time_steps`
- Can be implemented in a backwards compatible way by just throwing a warning, but copying `N` from `dims` to `opts`.

**Review by sandmaennchen — APPROVED** [review]
very cool :)

File: interfaces/acados_template/acados_template/acados_multiphase_ocp.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    All other phases are treated as intermediate phases, where only dynamics and path cost and constraints are used.
+
+    Limitations:
+    - state dimension must be the same for all phases (all other dimensions can be different), extending would require defining a transition function.
+    - solver options are shared between all phases. Interesting options to set phase-wise are:
+        - integrator_type
+        - collocation_type
+        - cost_discretization
+    """
+    def __init__(self, n_phases: int, N_list: list):
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply interfaces/acados_template/acados_template/acados_multiphase_ocp.py]: Done!

File: interfaces/acados_template/acados_template/acados_dims.py [hunk 2]
```diff
@@ -636,3 +637,9 @@ def N(self, N):
         else:
             raise Exception('Invalid N value, expected positive integer.')
 
+class MultiphaseOcpDims:
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply interfaces/acados_template/acados_template/acados_dims.py]: Done

**Pull request merged**